tests/main/interfaces-audio-playback-record/task.yaml

```diff
@@ -1,68 +1,58 @@
+systems: [ubuntu-16.04-*, ubuntu-18.04-*, ubuntu-19.10-*, ubuntu-20.04-*]
 
 environment:
     PLAY_FILE: "/snap/test-snapd-audio-record/current/usr/share/sounds/alsa/Noise.wav"
     # today, 16.04 and higher have a mediating pulseaudio
     EXFAIL: "ubuntu-14"
 
 prepare: |
-    # FIXME: This test is broken and should be ported to session-tool, systemd
-    # --user starts pulseaudio for us so we are always racing with it (whoever
-    # acquires the socket file wins).
-    #
-    # To prevent this from breaking this test mask pulseaudio.{socket,service}
-    # in all user sessions.
-    systemctl --user --global mask pulseaudio.{socket,service}
-
-    #shellcheck source=tests/lib/pkgdb.sh
-    . "$TESTSLIB"/pkgdb.sh
-    snap install --edge test-snapd-audio-record
+    truncate --size=0 defer.sh
+    chmod +x defer.sh
 
+    # Install pulseaudio.
     apt-get update
     apt-get install -y pulseaudio pulseaudio-utils
+    echo "apt-get autoremove --purge -y pulseaudio pulseaudio-utils" >>defer.sh
 
-    echo "Create XDG_RUNTIME_DIR=/run/user/12345"
-    # shellcheck disable=SC2174
-    mkdir -m 700 -p /run/user/12345 || true
-    chown test:test /run/user/12345
+    # Make sure the socket and the server is available in the user session.
+    if [ "$(systemctl --user --global is-enabled pulseaudio.socket)" != enabled ]; then
+        systemctl --user --global enable pulseaudio.socket
+        echo "systemctl --user --global disable pulseaudio.socket" >>defer.sh
+    fi
+    if [ "$(systemctl --user --global is-enabled pulseaudio.service)" != enabled ]; then
+        systemctl --user --global enable pulseaudio.service
+        echo "systemctl --user --global disable pulseaudio.service" >>defer.sh
+    fi
 
-    # ensure we have a clean pulse directory
-    test -d /home/test/.config && mv /home/test/.config /home/test/.config.spread
-    mkdir -m 700 /home/test/.config
-    mkdir -m 700 /home/test/.config/pulse
-    chown test:test /home/test/.config /home/test/.config/pulse
+    # Install a snap that uses the pulseaudio interface.
+    snap install --edge test-snapd-audio-record
+    echo "snap remove --purge test-snapd-audio-record" >>defer.sh
+
+    # Prepare a session for the user.
+    session-tool -u test --prepare
+    echo "session-tool -u test --restore" >>defer.sh
 
 restore: |
-    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "pulseaudio --kill" test || true
-    snap remove --purge test-snapd-audio-record
-    apt-get autoremove --purge -y pulseaudio pulseaudio-utils
-    rm -rf /run/user/12345 /home/test/.config/pulse
-    if [ -d /home/test/.config.spread ]; then
-        rm -rf /home/test/.config
-        mv /home/test/.config.spread /home/test/.config
-    fi
-    systemctl --user --global unmask pulseaudio.{socket,service}
+    # Restore system to the previous state by running deferred commands in
+    # reverse order.
+    tac defer.sh > refed.sh
```

Comment:
Did I mention how much I like this application of tac?

Reply:
Indeed, smart!